### PR TITLE
named datasets

### PR DESCRIPTION
**Pull request by ChristianBeilschmidt**
- [ ] I added an entry to [`CHANGELOG.md`](CHANGELOG.md) if knowledge of this change could be valuable to users.

---

Here is a brief summary of what I did:

<TEXT>

### CODE REVIEW

**Comment by ChristianBeilschmidt** [comment]
[Stable Names for datasets](https://trello.com/c/zJVzRWzc/852-stable-names-for-datasets)

File: datatypes/src/dataset.rs [hunk 1]
```diff
@@ -64,3 +61,297 @@ impl From<ExternalDataId> for DataId {
         DataId::External(value)
     }
 }
+
+pub const NAME_DELIMITER: char = ':';
+pub const SYSTEM_NAMESPACE: &str = "_";
+pub const SYSTEM_PROVIDER: &str = "_";
+
+/// The user-facing identifier for loadable data.
```
Comment (michaelmattig) [inline datatypes/src/dataset.rs:69]: shouldn't this be part of the services crate then?

File: datatypes/src/dataset.rs [hunk 2]
```diff
@@ -64,3 +61,297 @@ impl From<ExternalDataId> for DataId {
...
+        name: S2,
+    ) -> Self {
+        Self {
+            namespace: Self::canonicalize(namespace, SYSTEM_NAMESPACE),
+            provider: None,
+            name: name.into(),
+        }
+    }
+
+    pub fn with_global_provider<S1: Into<String> + PartialEq<&'static str>, S2: Into<String>>(
```
Comment (michaelmattig) [inline datatypes/src/dataset.rs:110]: can you add a comment what a global provider is?

File: datatypes/src/dataset.rs [hunk 3]
```diff
@@ -64,3 +61,297 @@ impl From<ExternalDataId> for DataId {
...
 }
+
+pub const NAME_DELIMITER: char = ':';
+pub const SYSTEM_NAMESPACE: &str = "_";
+pub const SYSTEM_PROVIDER: &str = "_";
+
+/// The user-facing identifier for loadable data.
+/// It can be resolved into a [`DataId`].
+#[derive(Debug, Clone, Hash, Eq, PartialEq)]
+pub struct NamedData {
```
Comment (michaelmattig) [inline datatypes/src/dataset.rs:72]: Can you maybe briefly explain how the named data and workspaces and providers work? Should also be in the docs then

File: datatypes/src/dataset.rs [hunk 4]
```diff
@@ -64,3 +61,297 @@ impl From<ExternalDataId> for DataId {
...
+        system_name: &'static str,
+    ) -> Option<String> {
+        if name == system_name {
+            None
+        } else {
+            Some(name.into())
+        }
+    }
+
+    pub fn with_global_name<S: Into<String>>(name: S) -> Self {
```
Comment (michaelmattig) [inline datatypes/src/dataset.rs:91]: is "global" and "system" data synonymous?

File: operators/src/engine/execution_context.rs [hunk 5]
```diff
@@ -59,6 +59,8 @@ pub trait ExecutionContext: Send
     async fn read_ml_model(&self, path: PathBuf) -> Result<String>;
 
     async fn write_ml_model(&mut self, path: PathBuf, ml_model_str: String) -> Result<()>;
+
+    async fn resolve_named_data(&self, data: &NamedData) -> Result<DataId>;
```
Comment (michaelmattig) [inline operators/src/engine/execution_context.rs:63]: as discussed, I think we can remove this method and have the `MetaDataProvider`s `meta_data` method just accept a `NamedData` directly, if it is able to resolve it anyway.

On the other hand it would also be neat, if the `NamedData` was only a thing in the services crate and the other crates just worked wit `DataId`s but I guess this is not really feasible?

File: services/src/handlers/workflows.rs [hunk 6]
```diff
@@ -677,6 +675,7 @@ mod tests {
     use actix_web::{http::header, http::Method, test};
     use actix_web_httpauth::headers::authorization::Bearer;
     use geoengine_datatypes::collections::MultiPointCollection;
+    use geoengine_datatypes::dataset::NamedData;
```
Comment (michaelmattig) [inline services/src/handlers/workflows.rs:678]: not sure where to put this comment but: on GET /workflow the system namespace should be omitted as well, or not?

File: services/src/datasets/listing.rs [hunk 7]
```diff
@@ -78,6 +79,8 @@ pub trait DatasetProvider: Send
     async fn load_dataset(&self, dataset: &DatasetId) -> Result<Dataset>;
 
     async fn load_provenance(&self, dataset: &DatasetId) -> Result<ProvenanceOutput>;
+
+    async fn resolve_dataset(&self, name: &DatasetName) -> Result<DatasetId>;
```
Comment (michaelmattig) [inline services/src/datasets/listing.rs:83]: resolve_dataset_id maybe?

File: services/src/handlers/wfs.rs [hunk 8]
```diff
@@ -1180,14 +1180,14 @@ x;y
                     "vector": {
                         "type": "OgrSource",
                         "params": {
-                            "data": ne_10m_ports_id,
+                            "data": "points_with_time",
```
Comment (michaelmattig) [inline services/src/handlers/wfs.rs:1183]: is it a good idea to hardcode this her? What if we want to change it in the future? Then we'd have to change it everywhere.

File: services/src/pro/datasets/in_memory.rs [hunk 9]
```diff
@@ -902,8 +954,8 @@ mod tests {
         };
 
         let ds = AddDataset {
-            id: None,
-            name: "OgrDataset".to_string(),
+            name: None,
```
Comment (michaelmattig) [inline services/src/pro/datasets/in_memory.rs:957]: can you add a test where the name is not None?

File: services/src/apidoc.rs [hunk 10]
```diff
@@ -128,6 +128,8 @@ use utoipa::{Modify, OpenApi};
             DataId,
             DataProviderId,
             DatasetId,
+            DatasetName,
+            NamedData,
```
Comment (michaelmattig) [inline services/src/apidoc.rs:132]: You need to adjust the pro apidoc.rs too

File: test_data/provider_defs/mock.json [hunk 11]
```diff
@@ -27,11 +27,7 @@
           "operator": {
             "type": "MockDatasetDataSource",
             "params": {
-              "data": {
-                "type": "external",
-                "providerId": "d0535f1d-27b6-4982-b2f8-b1070f1bf6ee",
-                "layerId": "myMock"
-              }
+              "data": "_:d0535f1d-27b6-4982-b2f8-b1070f1bf6ee:myMock"
```
Comment (michaelmattig) [inline test_data/provider_defs/mock.json:30]: can the "_" be omitted?

File: datatypes/src/dataset.rs (hunk 1 above)
Comment (ChristianBeilschmidt) [reply datatypes/src/dataset.rs:69]: No, this should be part of the place where the API resides, since it translates `_` to None.
Since the API is at datatypes as well as the services crate, which is a bad design, we have to leave it there.

File: operators/src/engine/execution_context.rs (hunk 5 above)
Comment (ChristianBeilschmidt) [reply operators/src/engine/execution_context.rs:63]: We cannot do this, since it is also part of operators, which are also part of our API.
We should work on our API encapsulation. Then, we could do this.

File: services/src/handlers/workflows.rs (hunk 6 above)
Comment (ChristianBeilschmidt) [reply services/src/handlers/workflows.rs:678]: Is this part of an external provider or is this the internal one?

File: services/src/handlers/wfs.rs (hunk 8 above)
Comment (ChristianBeilschmidt) [reply services/src/handlers/wfs.rs:1183]: Do you have an idea on not to hardcode this? I could change it, then.

File: services/src/apidoc.rs (hunk 10 above)
Comment (ChristianBeilschmidt) [reply services/src/apidoc.rs:132]: Didn't I do this? Isn't there a CI check for this?

File: datatypes/src/dataset.rs (hunk 2 above)
Comment (ChristianBeilschmidt) [reply datatypes/src/dataset.rs]: [6e97ae8](https://github.com/geo-engine/geoengine/pull/782/commits/6e97ae81f0f45f0fd46f095e57ae0b6238f9e15d)

File: datatypes/src/dataset.rs (hunk 3 above)
Comment (ChristianBeilschmidt) [reply datatypes/src/dataset.rs:82]: [6e97ae8](https://github.com/geo-engine/geoengine/pull/782/commits/6e97ae81f0f45f0fd46f095e57ae0b6238f9e15d)

File: datatypes/src/dataset.rs (hunk 4 above)
Comment (ChristianBeilschmidt) [reply datatypes/src/dataset.rs]: [6e97ae8](https://github.com/geo-engine/geoengine/pull/782/commits/6e97ae81f0f45f0fd46f095e57ae0b6238f9e15d)

File: services/src/datasets/listing.rs (hunk 7 above)
Comment (ChristianBeilschmidt) [reply services/src/datasets/listing.rs:83]: 9e737b1: resolve dataset to id naming

File: services/tests/drivers.rs [hunk 12]
```diff
@@ -0,0 +1,31 @@
+//! We needto run this in a separate process since it changes the global state of the GDAL driver
```
[the remark this answers is not in the document]
Comment (ChristianBeilschmidt) [reply services/tests/drivers.rs:1]: https://github.com/geo-engine/geoengine/commit/63a3f5fbc55bbb0a0ce7b52110bc63c0aaf7543b

File: services/src/pro/datasets/in_memory.rs (hunk 9 above)
Comment (ChristianBeilschmidt) [reply services/src/pro/datasets/in_memory.rs:944]: https://github.com/geo-engine/geoengine/commit/6f195958364afd368eaed45db4d3c01d83cbd2e7

File: services/src/handlers/workflows.rs (hunk 6 above)
Comment (michaelmattig) [reply services/src/handlers/workflows.rs:678]: it's the mock provider. I guess we can't omit the system namespace because then we wouldn't know whether the uuid is the namespace or the provider right?

File: services/src/apidoc.rs (hunk 10 above)
Comment (michaelmattig) [reply services/src/apidoc.rs:135]: no you didn't. At least I cannot find `NamedData` in src/pro/apidoc.rs

**Review by michaelmattig — APPROVED** [review]
